### PR TITLE
Tiny containers multi stage

### PR DESCRIPTION
**Pull request by shaunsmith**
Moved all building into multi-stage Docker builds to remove need to setup Musl and UPX.
Also added three JVM examples for contrast.

### CODE REVIEW

File: tiny-java-containers/README.md [hunk 1]
```diff
@@ -22,119 +22,70 @@ App](images/youtube.png)](https://youtu.be/6wYrAtngIVo)
...
-
-Clone this Git repo and in your Linux shell type the following to download and
-configure the `musl` toolchain.
-
-![](images/keyboard.jpg) `./setup-musl.sh`
-
-Download [upx](https://upx.github.io/):
-
-![](images/keyboard.jpg) `./setup-upx.sh`
+Clone this Git repo.  Everything runs in Docker so no need to install anything
```
Comment (olyagpl) [inline tiny-java-containers/README.md:30]:
```suggestion
Clone this Git repo. Everything runs in Docker so no need to install anything
```

File: tiny-java-containers/README.md [hunk 2]
```diff
@@ -22,119 +22,70 @@ App](images/youtube.png)](https://youtu.be/6wYrAtngIVo)
...
+on your machine.
 
 ## Hello World
 
-With the `musl` toolchain installed, cd in to the `helloworld` folder.
+Let's start with a simple Hello World example.
+
+CD into the `helloworld` folder.
```
Comment (olyagpl) [inline tiny-java-containers/README.md:37]:
```suggestion
Change the directory to `helloworld`.
```

File: tiny-java-containers/README.md [hunk 3]
```diff
@@ -22,119 +22,70 @@ App](images/youtube.png)](https://youtu.be/6wYrAtngIVo)
...
-   start extremely fast as there is minimal startup cost.
-
-3. The `upx` compressed executable is over 70% smaller, 1.7MB vs. 6.3MB! With
-   `upx` the application self-extracts quickly but does incur a cost of about
-   100ms for decompression. See this blog for a deep dive on [GraalVM Native
-   Image and
+1. The `hello` executable generated by GraalVM Native Image in the Dockerfile
+   using the `--static --libc=musl` options is a fully self-contained
+   executable. This means that it does not rely on any libraries in the host
+   operating system environment.  This makes it easier to package in a variety
```
Comment (olyagpl) [inline tiny-java-containers/README.md:59]:
```suggestion
   operating system environment. This makes it easier to package in a variety of container images.
```

File: tiny-java-containers/helloworld/Dockerfile [hunk 4]
```diff
@@ -1,3 +1,29 @@
...
+ARG UPX_VERSION=4.2.2
+ARG UPX_ARCHIVE=upx-${UPX_VERSION}-amd64_linux.tar.xz
+RUN microdnf -y install wget xz && \
+    wget -q https://github.com/upx/upx/releases/download/v${UPX_VERSION}/${UPX_ARCHIVE} && \
+    tar -xJf ${UPX_ARCHIVE} && \
+    rm -rf ${UPX_ARCHIVE} && \
+    mv upx-${UPX_VERSION}-amd64_linux/upx . && \
+    rm -rf upx-${UPX_VERSION}-amd64_linux
+
+    # Compile the Hello class to Java bytecode
```
Comment (olyagpl) [inline tiny-java-containers/helloworld/Dockerfile:15]:
```suggestion
# Compile the Hello class to Java bytecode
```

File: tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.dynamic [hunk 5]
```diff
@@ -1,5 +1,11 @@
+FROM container-registry.oracle.com/graalvm/native-image:23 AS nativebuild
+WORKDIR /build
+# Build a dynamically linked native executable optimizing for size with `-Os`
+RUN native-image -Os --enable-sbom=cyclonedx -m jdk.httpserver -o jwebserver.dynamic
```
Comment (olyagpl) [inline tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.dynamic:4]: We don't explain the `--enable-sbom=cyclonedx ` option in the text. Do you need to add it also to the full static Dockerfile build?

File: tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.jlink [hunk 6]
```diff
@@ -1,5 +1,21 @@
+FROM container-registry.oracle.com/graalvm/jdk:23 AS build
+RUN microdnf install -y binutils
```
Comment (olyagpl) [inline tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.jlink:2]: Why do we need to install `binutils`?

**Review by olyagpl — CHANGES_REQUESTED** [review]

File: tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.dynamic (hunk 5 above)
Comment (shaunsmith) [reply tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.dynamic:4]: Oh, I'll remove that for now.

**Review by fniephaus — DISMISSED** [review]
LGTM

File: tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.jlink (hunk 6 above)
Comment (shaunsmith) [reply tiny-java-containers/jwebserver/Dockerfile.distroless-java-base.jlink]: Must have been needed earlier and not now.  Removed.

**Review by olyagpl — APPROVED** [review]

**Pull request merged**